Support specifying IPv6 proxies in CIDR notation

sleiner opened the pull request:
Previously, [it was not possible to use CIDR notation for IPv6 proxies in the trusted_proxies parameter of config.php][1]. This patch adds support for that, mainly by relying on inet_pton and a custom bitwise string comparison function instead of ip2long (which does not work for IPv6 because of its 128 bit address space).

It is worth noting that the bitwise comparison could be implemented a bit more straightforward using [GMP][2], but since Nextcloud does not currently require GMP to be available, this alternative implementation was chosen instead.

The patch fixes issue #32253.

[1]: https://docs.nextcloud.com/server/24/admin_manual/configuration_server/reverse_proxy_configuration.html#defining-trusted-proxies
[2]: https://www.php.net/manual/en/class.gmp.php

sleiner commented:
Hi @kesselb,

I had not seen #12535 before, thanks for pointing me to that! 😊 After taking a brief look at it, here are my thoughts:

 - The test cases for the `Request` class are basically the same as far as I see (except for some some of the concrete IP addresses and the extra constructor parameters introduced in #12535). IMO, that's a a good sign.
 - Compared to #12535, this implementation is much simpler:
   - No new classes, just two more helper methods
   - No custom-built detection of IPv4 vs. IPv6 addresses, just the use of `inet_pton`
   - No regexes validating IP addresses, just the use of `inet_pton`
 - Still, there are two places were this patch implements custom logic:
   - **Splitting a string possibly containing a CIDR representation up into the net address and the prefix length**: This just splits the string by the `/` character if there is one (same as [Symfony's IpUtils](https://github.com/symfony/http-foundation/blob/18f8a2a3ab703428143c27c055fd743ab7e7dcb1/IpUtils.php#L71)).
   - **Detecting whether two packed strings have the same prefix with a variable length of bits**: Since I did not want to rely on GMP being available, this involves some custom fiddling with bits. Symfony's IpUtils do the same for IPv6, although [their implementation](https://github.com/symfony/http-foundation/blob/18f8a2a3ab703428143c27c055fd743ab7e7dcb1/IpUtils.php#L138) differs quite a bit from mine.
 - Compared to my implementation, the Symfony has a few advantages:
   - It has been used in production before.
   - It _may_ perform better: For my implementation, I can say that I have not benchmarked it. In any case, the symfony implementation uses a cache while mine does not. This may be beneficial.

In general, I have no strong attachment to the specific code of this patch, I mainly would like to use the feature ;-) I have never contributed to Nextcloud (or written any PHP, for that matter) before so I had no idea what your policies are concerning pulling in code from external libraries. So just writing the code myself seemed like the quickest way to me. So let me know what you and your colleagues consider the best way forward here 👍🏻

sleiner commented on lib/private/AppFramework/Http/Request.php:
```diff
@@ -574,25 +575,86 @@ public function getId(): string {
 	protected function matchesTrustedProxy($trustedProxy, $remoteAddress) {
-		$cidrre = '/^([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})\/([0-9]{1,2})$/';
+		[$proxyAddress, $prefixLength] = array_pad(explode('/', $trustedProxy, 2), 2, null);
+		return $this->ipAddressInNetwork(inet_pton($remoteAddress),
+										 inet_pton($proxyAddress),
```
If `$remoteAddress` or `$proxyAddress` are not valid IP addresses, `inet_pton` will return `false` and a `TypeError` will be thrown.

That case must be handled gracefully.

sleiner replied:
Fixed in 74c5cff4febf2f4194ce6e2b96620cea0b83c51c.

sleiner commented:
@kesselb could you please approve running the workflows again? 😅

sleiner commented:
@kesselb Can you give an ETA for the reviews?

skjnldsv reviewed (approved):
Code looks good

come-nc commented:
@sleiner Hello, in the end we added symfony/http-foundation as a dependency because we needed the same functionnality in other parts of the code and felt it was better to rely on symfony rather than maintain our own copy.
Could you refactor this PR to use Symfony\Component\HttpFoundation\IpUtils instead? (rebase on master first)

I could do that in an other PR but I think your tests can work the same and this is valuable.

sleiner commented:
Sure thing 👍🏻 (though it will probably take a few days before I'll get to it)

sleiner committed:
Support specifying IPv6 proxies in CIDR notation

Previously, it was not possible to use CIDR notation for IPv6 proxies
in the trusted_proxies parameter of config.php [1]. This patch adds
support for that.

[1]: https://docs.nextcloud.com/server/24/admin_manual/configuration_server/reverse_proxy_configuration.html#defining-trusted-proxies

Signed-off-by: Simon Leiner <simon@leiner.me>

sleiner commented:
@come-nc done 😊

come-nc commented:
@sleiner Can you rebase on master to retrigger CI?
It feels like the failures are unrelated.

sleiner commented:
@come-nc done 👍🏻

sleiner commented:
we still have some failing runs :/

- The failing performance test seems like a general issue with pull requests from forked repositories to me.
- The drone build fails on a timeout (for a notification in the files app), so this is probably a flaky test?

come-nc commented:
> we still have some failing runs :/
> 
> * The failing performance test seems like a general issue with pull requests from forked repositories to me.
> * The drone build fails on a timeout (for a notification in the files app), so this is probably a flaky test?

Yeah seems unrelated, I merged it.
